### DISCUSSION TITLE
Can I replicate nginx's "time_iso8601" variable in unit?

### DISCUSSION THREAD

**Discussion by bunny-therapist** [nginx/unit#1118] [Support] [answered]
I am experimenting with porting an nginx+uwsgi setup to unit, and when trying to replicate the access log format, I ran into the problem that the nginx format used the `time_iso8601` variable (described [here](http://nginx.org/en/docs/http/ngx_http_core_module.html)), but I could not find such a variable in unit, just `time_local`. Is there another way to get ISO 8601 format for the time?

**Comment by lcrilly** [comment]
You can get an ISO date string with a JavaScript template literal. Here is a sample configuration for a very simple log format.

```json
{
  "access_log": {
    "path": "/dev/stdout",
    "format": "`@timestamp=${new Date().toISOString()} ip=${remoteAddr} uri=${uri}`"
  }
}
```

However, when using JavaScript, you don't currently have access to the rest of Unit's variables (issue https://github.com/nginx/unit/issues/977) until (PR https://github.com/nginx/unit/pull/1024) arrives in the next release.

**Reply by bunny-therapist** [reply]
I am not sure what this "key=value" structure represents. Is that the raw shape of the log text? I want the log output to be in json format, so I wrote the whole log string as a json string. I tried using the "new Date()..." code there, but it did not work. Was that just because I used ot

This is my current log format that I want to have an ISO timestamp in:
```
"{\"@timestamp\":\"$time_local\",\"remote_addr\":\"$remote_addr\",\"request\":\"$request_line\",\"status\": \"$status\",\"body_bytes_sent\":\"$body_bytes_sent\",\"request_time\":\"$request_time\",\"http_referrer\":\"$header_referer\",\"http_user_agent\":\"$header_user_agent\", \"request_id\":\"$response_header_x_request_id\"}"
```

This is the old nginx log config that I am trying to replicate:
```
log_format json_combined
  '{"@timestamp":"$time_iso8601","remote_addr":"$remote_addr","remote_user":"$remote_user","request":"$request","status": "$status","body_bytes_sent":"$body_bytes_sent","request_time":"$request_time","http_referrer":"$http_referer","http_user_agent":"$http_user_agent", "request_id":"$sent_http_x_request_id"}';
  access_log /dev/stdout json_combined
```

Will I be able to replicate this somehow?

**Reply by lcrilly** [reply, marked as answer]
This is as close as I can get with Unit 1.31.1
```json
"format": "`{\"@timestamp\":\"${new Date().toISOString()}\",\"remote_addr\":\"${remoteAddr}\",\"request\":\"${uri}\",\"http_referer\":\"${headers['Referer'] === undefined ? '-' : headers['Referer']}\",\"http_user_agent\":\"${headers['User-Agent']}\"}`"
```
Note that request is just the URI and that status, bytes, request_time are missing.
```
$ curl -i localhost:9000
HTTP/1.1 204 No Content
Server: Unit/1.31.1
Date: Thu, 08 Feb 2024 15:11:33 GMT
Content-Length: 0

{"@timestamp":"2024-02-08T15:11:33.616Z","remote_addr":"127.0.0.1","request":"/","http_referer":"-","http_user_agent":"curl/8.4.0"}
```

As I mentioned, JavaScript doesn't currently allow you to access all of Unit's variables. That's coming in 1.32.

However, it does allow you to do things you cannot with nginx, such as log all of the request headers
```json
"format": "`{\"timestamp\":\"${new Date().toISOString()}\",\"connection\":{\"client_ip\":\"${remoteAddr}\"},\"request\":{\"uri\":\"${uri}\",\"headers\":${(() => { var h = {}; Object.keys(headers).map(k => h[k] = headers[k]); return JSON.stringify(h); } )()}}}\n`"
```

**Reply by bunny-therapist** [reply]
Thank you!
